sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs

```diff
@@ -0,0 +1,309 @@
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException("request");
+            }
+
+            if (_mode == LoginMode.Basic)
+            {
+                request.Headers.Authorization = new AuthenticationHeaderValue("Basic", _authHeader);
+            }
+            else
+            {
+                string operation = "https://" + _loginUrl + request.RequestUri.AbsolutePath;
+                string scope = await getScope(operation, request.Method.Method, request.RequestUri.AbsolutePath);
+
+                request.Headers.TryAddWithoutValidation("Authorization", "Bearer " + getAcrAccessToken(scope));
+            }
+            await base.ProcessHttpRequestAsync(request, cancellationToken);
+
+        }
+
+        #endregion
+
+        #region Helpers
+
+        /// <summary>
+        /// Acquires a new ACR access token if necessary. It can also acquire a cached access token in order to avoid extra requests to
+        /// the oauth2 endpoint improving efficiency.
+        /// <param name='scope'> The scope for the particuar operation. Can be obtained from the Www-Authenticate header.
+        /// </summary>
+        public string getAcrAccessToken(string scope)
+        {
+            if (_mode == LoginMode.Basic)
+            {
+                throw new Exception("This Function cannot be invoked for requested Login Mode. Basic Authentication does not support JWT Tokens ");
+            }
+
+            // If Token is available or existed and can be renewed
+            if (_acrAccessTokens.ContainsKey(scope))
+            {
+                if (!_acrAccessTokens[scope].CheckAndRefresh())
+                {
+                    throw new Exception("Access Token for scope " + scope + " expired and could not be refreshed");
+                }
+
+                return _acrAccessTokens[scope].Value;
+            }
+
+            if (_mode == LoginMode.TokenAad)
+            {
+                _acrAccessTokens[scope] = new AcrAccessToken(_acrRefresh, scope, _loginUrl);
+            }
+            else if (_mode == LoginMode.TokenAuth)
+            {
+                _acrAccessTokens[scope] = new AcrAccessToken(_username, _password, scope, _loginUrl);
+            }
+
+            return _acrAccessTokens[scope].Value;
+        }
+
+        /// <summary>
+        /// Acquires the required scope for a specific operation. This will be done by obtaining a challenge and parsing out the scope
+        /// from the ww-Authenticate header. In the event of failure (Some endpoints do not seem to return the scope) it will attempt
+        /// resolution through a small local resolver.
+        /// <param name='scope'> The scope for the particuar operation. Can be obtained from the Www-Authenticate header.
+        /// </summary>
+
+        public async Task<string> getScope(string operation, string method, string path)
+        {
+
+            if (_acrScopes.ContainsKey(method + ">" + operation))
```

Comment:
Suggestion:

Consider changing this to:

```csharp
string methodAndOperation = $"{method}>{operation}";
if (_arcScopes.TryGetValue(methodAndOperation, out string result))
{
   return result;
}
```

That reduces the string allocations as well as reduces the calls into the Dictionary methods.

Reply:
Thanks! Will do